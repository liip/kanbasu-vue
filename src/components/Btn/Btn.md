```vue
<btn variant="default">
  Hello
</btn>
```

### Block button

```vue
<btn variant="primary" block>
  Hello
</btn>
```

### Large primary button as anchor element

```vue
<btn tag="a" href="#" variant="primary" size="large">
  Hello
</btn>
```

### Bare disabled submit button

```vue
<btn variant="bare" type="submit" disabled>
  Hello
</btn>
```