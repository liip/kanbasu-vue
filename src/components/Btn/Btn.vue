<template>
  <component
    :is="tag"
    :class="classNames"
    v-bind="attributes"
    v-on="$listeners"
  >
    <slot></slot>
  </component>
</template>

<script>
export default {
  name: 'Btn',

  inheritAttrs: false,

  props: {
    /**
     * The HTML tag used for the button.
     */
    tag: {
      type: String,
      default: 'button',
    },
    /**
     * The `type` attribute for input and button tags
     */
    type: {
      type: String,
      default: 'button',
    },
    /**
     * The style of the button:
     * `default, primary, bare`
     */
    variant: {
      type: String,
    },
    /**
     * The button size:
     * `small, large`
     */
    size: {
      type: String,
    },
    /**
     * Make the button takes 100% of the parent width
     */
    block: {
      type: Boolean,
      default: false,
    },
    /**
     * Set the disable attribute and add the `disabledClass` to the button
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * The class applied to disabled buttons
     */
    disabledClass: {
      type: String,
      default: 'btn--disabled',
    },
  },

  computed: {
    attributes() {
      return Object.assign(
        {},
        {
          // Set disabled attribute only on button/input tags
          ...(this.disabled &&
            ['input', 'button'].indexOf(this.tag) > -1 && { disabled: true }),
          // Set type attribute only on button/input tags and defaults to `button`
          ...(this.type &&
            ['input', 'button'].indexOf(this.tag) > -1 && { type: this.type }),
        },
        this.$attrs
      );
    },

    classNames() {
      return [
        'btn',
        { 'btn--block': this.block },
        { [`btn--${this.variant}`]: !!this.variant },
        { [`btn--${this.size}`]: !!this.size },
        { [this.disabledClass]: this.disabled },
      ];
    },
  },
};
</script>
