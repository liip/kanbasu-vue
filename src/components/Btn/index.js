import Btn from './Btn.vue';

Btn.install = function(Vue) {
  Vue.component(Btn.name, Btn);
};

export default Btn;
