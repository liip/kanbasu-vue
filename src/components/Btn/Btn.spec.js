import { shallowMount } from '@vue/test-utils';
import Btn from './Btn.vue';

describe('Btn.vue', () => {
  it('renders as button with type by default', () => {
    const wrapper = shallowMount(Btn);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as input with type submit', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        tag: 'input',
        type: 'submit',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as anchor without type', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        tag: 'a',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with variant', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        variant: 'primary',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with size', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        size: 'small',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as block', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        block: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as disabled', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        disabled: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders as disabled with custom class', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        disabled: true,
        disabledClass: 'disabled',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without disabled attribute on anchors', () => {
    const wrapper = shallowMount(Btn, {
      propsData: {
        tag: 'a',
        disabled: true,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
