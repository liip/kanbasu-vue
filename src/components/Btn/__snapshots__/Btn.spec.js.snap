// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Btn.vue renders as anchor without type 1`] = `<a class="btn"></a>`;

exports[`Btn.vue renders as block 1`] = `<button type="button" class="btn btn--block"></button>`;

exports[`Btn.vue renders as button with type by default 1`] = `<button type="button" class="btn"></button>`;

exports[`Btn.vue renders as disabled 1`] = `<button disabled="disabled" type="button" class="btn btn--disabled"></button>`;

exports[`Btn.vue renders as disabled with custom class 1`] = `<button disabled="disabled" type="button" class="btn disabled"></button>`;

exports[`Btn.vue renders as input with type submit 1`] = `<input type="submit" class="btn">`;

exports[`Btn.vue renders with size 1`] = `<button type="button" class="btn btn--small"></button>`;

exports[`Btn.vue renders with variant 1`] = `<button type="button" class="btn btn--primary"></button>`;

exports[`Btn.vue renders without disabled attribute on anchors 1`] = `<a class="btn btn--disabled"></a>`;
